src/high-level/abstract-tensor.lisp

```diff
@@ -0,0 +1,262 @@
+;;;; abstract-tensor.lisp
+;;;;
+;;;; Author: Cole Scott
+;;;;
+;;;; Collaborators: Robert Smith
+
+(in-package #:magicl)
+
+(defstruct (abstract-tensor
+            (:constructor nil)
+            (:copier nil))
+  "Abstract tensor class. Superclass for implementing the abstract-tensor protocol")
+
+;;; abstract-tensor protocol
+;;; These methods do not have generic definitions and must be implemented by subclasses
+
+(defgeneric shape (tensor)
+  (:documentation "The shape (dimensions) of the tensor. eg. '(2 3) for a 2x3 tensor"))
+
+(defgeneric tref (tensor &rest pos)
+  (:documentation "Get a reference to element at position"))
+
+(defgeneric (setf tref) (new-value tensor &rest pos)
+  (:documentation "Set the value of element at position"))
+
+(defgeneric copy-tensor (tensor &rest args)
+  (:documentation "Create a new tensor with the same properties as the given tensor, creating new storage without initializing contents"))
+
+(defgeneric deep-copy-tensor (tensor &rest args)
+  (:documentation "Create a new tensor with the same properties as the given tensor, copying the contents of the storage"))
+
+;;; abstract-tensor generic methods
+;;; These methods have default generic implementations but can be optimized by subclasses
+
+(defgeneric order (tensor)
+  (:documentation "Order (number of dimensions) of the tensor")
+  (:method ((tensor abstract-tensor))
+    (length (shape tensor))))
+
+(defgeneric size (tensor)
+  (:documentation "Total number of elements in the tensor")
+  (:method ((tensor abstract-tensor))
+    (reduce #'* (shape tensor))))
+
+(defgeneric element-type (tensor)
+  (:documentation "The type of the elements in the tensor")
+  (:method ((tensor abstract-tensor))
+    (declare (ignore tensor))
+    t))
+
+(defgeneric lisp-array (tensor &optional target)
+  (:documentation "Return a lisp array containing the data from the tensor
+
+If TARGET is specified then the contents of the tensor are copied into the array.
+In the event TARGET is not specified, the result may return an array sharing memory with the input tensor.")
+  (:method :before (tensor &optional target)
+    (unless (null target)
+      (policy-cond:policy-if
+       (< speed safety)
+       (assert (and (= (order tensor) (array-rank target))
+                    (equal (shape tensor) (array-dimensions target))))
+       nil)))
```

Comment:
Nitpick: this is screaming to be a `policy-when` instead (if it existed :-). Also, if you stick with `policy-if` please move the condition to the previous line.

Ditto on all of the instances of `policy-if` in this file.

Reply:
Changed to `with-expectation`